java-frontend/src/main/java/org/sonar/java/model/VisitorsBridge.java

```diff
@@ -124,12 +132,74 @@ public void visitFile(@Nullable Tree parsedTree) {
       runScanner(javaFileScannerContext, new SymbolicExecutionVisitor(executableScanners, behaviorCache), AnalysisError.Kind.SE_ERROR);
       behaviorCache.cleanup();
     }
-    executableScanners.forEach(scanner -> runScanner(javaFileScannerContext, scanner, AnalysisError.Kind.CHECK_ERROR));
+    scannerRunner.run(javaFileScannerContext);
+    executableScanners.stream().filter(s -> s instanceof SymbolicExecutionVisitor || !(s instanceof IssuableSubscriptionVisitor))
+      .forEach(scanner -> runScanner(javaFileScannerContext, scanner, AnalysisError.Kind.CHECK_ERROR));
     if (semanticModel != null) {
       classesNotFound.addAll(semanticModel.classesNotFound());
     }
   }
 
+  private static class ScannerRunner {
+    private EnumMap<Tree.Kind, List<SubscriptionVisitor>> checks;
+    private List<SubscriptionVisitor> subscriptionVisitors;
+
+    public ScannerRunner(List<JavaFileScanner> executableScanners) {
+      checks = new EnumMap<>(Tree.Kind.class);
+      subscriptionVisitors = executableScanners.stream()
+        .filter(s -> s instanceof IssuableSubscriptionVisitor)
+        .map(s -> (SubscriptionVisitor) s).collect(Collectors.toList());
+      subscriptionVisitors.forEach(s -> s.nodesToVisit().forEach(k -> {
+        if (!checks.containsKey(k)) {
+          checks.put(k, new ArrayList<>());
+        }
+        checks.get(k).add(s);
+      })
+      );
+    }
+
+
+    public void run(JavaFileScannerContext javaFileScannerContext) {
+      subscriptionVisitors.forEach(s -> s.setContext(javaFileScannerContext));
+      visit(javaFileScannerContext.getTree());
+      subscriptionVisitors.forEach(s -> s.leaveFile(javaFileScannerContext));
```

Comment:
There is something I really don't like here... But I have no argument strictly against this approach. I'm however just wondering about the memory impact.

 We have numerous checks which are keeping in memory (references to) some nodes (for hierarchy checking for instance, reporting, or whatever other cases). Previously, this memory was (theoretically) systematically cleaned when leaving each file, and filled up again with a new check. With this new approach, the memory usage of each node is going to grow as long as we didn't reach the end of file. Do we want to live with this? On the long run it may means asking for more memory for an analysis.

Reply:
As discussed this is not a lot different from what we have today when keeping a context within a file : most of the checks are only clearing their references to AST when entering the next file anyway.